ops/lib.rs

```diff
@@ -300,6 +339,102 @@ fn codegen_v8_sync(
   }
 }
 
+fn can_be_fast_api(
+  core: &TokenStream2,
+  f: &syn::ItemFn,
+) -> Option<(TokenStream2, TokenStream2)> {
+  let inputs = &f.sig.inputs;
+  let output = &f.sig.output;
+
+  let ret = match is_fast_scalar(core, output, true) {
+    Some(ret) => ret,
+    None => return None,
+  };
+
+  // need a reciever for sequence types.
+  let mut needs_recv = false;
+  let mut args = vec![];
+  for input in inputs {
+    let ty = match input {
+      syn::FnArg::Typed(pat) => &pat.ty,
+      _ => unreachable!(),
+    };
+    match is_fast_scalar(core, ty, false) {
+      None => match is_fast_arg_sequence(core, ty) {
+        Some(arg) => {
+          needs_recv = true;
+          args.push(arg);
+        }
+        None => return None,
+      },
+      Some(arg) => {
+        args.push(arg);
+      }
+    }
+  }
+
+  let args = args
+    .iter()
+    .fold(quote! {}, |acc, new| quote! { #acc, #new });
+  Some((args, ret))
+}
+
+// A v8::Local<v8::Array> or FastApiTypedArray<T>
+fn is_fast_arg_sequence(
+  core: &TokenStream2,
+  ty: impl ToTokens,
+) -> Option<TokenStream2> {
+  if is_fast_typed_array(&ty) {
+    return Some(
+      quote! { #core::v8::fast_api::Type::TypedArray(#core::v8::fast_api::CType::Uint32) },
+    );
+  }
+  if is_local_array(&ty) {
+    return Some(
+      quote! { #core::v8::fast_api::Type::Sequence(#core::v8::fast_api::CType::Void) },
+    );
+  }
+  None
+}
+
+fn is_local_array(arg: impl ToTokens) -> bool {
+  static RE: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"^v8::Local<v8::Array>$").unwrap());
+  RE.is_match(&tokens(arg))
+}
+
+fn is_fast_typed_array(arg: impl ToTokens) -> bool {
+  static RE: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(r#": (?:deno_core :: )?FastApiTypedArray$"#).unwrap()
+  });
+  RE.is_match(&tokens(arg))
+}
+
+fn is_fast_scalar(
+  core: &TokenStream2,
+  ty: impl ToTokens,
+  is_ret: bool,
+) -> Option<TokenStream2> {
+  let cty = if is_ret {
+    quote! { CType }
+  } else {
+    quote! { Type }
+  };
+  if is_resource_id(&ty) {
+    return Some(quote! { #core::v8::fast_api::#cty::Uint32 });
+  }
+  if is_void(&ty) {
+    return Some(quote! { #core::v8::fast_api::#cty::Void });
+  }
+  match tokens(&ty).as_str() {
+    "u8" | "u16" | "u32" => Some(quote! { #core::v8::fast_api::#cty::Uint32 }),
```

Comment:
Okay, I did some sleuthing into this and it seems that indeed using a smaller width return value as a larger width value is not safe. When u8 / u16 values are passed as an argument, the caller is expected to zero out the bits up to in the lower half of the register. (Actually, this might mean that even passing a u32 off as a u8 may cause a good degree of UB if the value actually overflows the u8 max limit. Users beware.)

No such guarantee (well, even the sign / zero extending is an undocumented convention that clang relies on) exists for return values, and they can contain garbage beyond their width.

So, when a native function returns an u8 containing `0b00000001`, the register might actually contain `0b11111111111111111111111100000001` and naturally reading that as a u32 (as V8 will do) will give `4294967041` instead of the expected `1`.

Some compilers might still do zero / sign bit extending on return values, but it's not a given across all systems. Also, negative sign bit extending will of course also cause the value to become quite odd indeed.

Reply:
Thinking on this a tad bit more:
1. For parameters, this means that the responsibility of upholding the ABI contract is transferred to the JS code, ie. essentially to whoever is using Deno FFI be that a library developer or a user of said library if raw bindings are exposed.
2. For return values, again this means that the Deno FFI user must simply expect u8 / i8 / u16 / i16 return values to potentially contain values beyond the expected value and thus they must manually zero out the top bits of any return values.